src/vmm/src/vstate/vcpu/mod.rs

```diff
@@ -624,10 +647,136 @@ mod tests {
+            if (&*self.vcpu_exit_reason.lock().unwrap()).is_some() {
+                return self.vcpu_exit_reason.lock().unwrap().take().unwrap();
+            }
+            Err(errno::Error::new(libc::SIGILL))
+        }
+    }
+
+    #[test]
+    fn test_run_emulation() {
+        let (_vm, mut vcpu, _vm_mem) = setup_vcpu(0x1000);
+        vcpu.vcpu_exit_reason = Mutex::new(Some(Ok(VcpuExit::Hlt)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(res.unwrap(), VcpuEmulation::Stopped);
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::Shutdown));
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(res.unwrap(), VcpuEmulation::Stopped);
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::FailEntry));
+        let res = vcpu.run_emulation();
+        assert!(res.is_err());
+        assert_eq!(
+            format!("{:?}", res.unwrap_err()),
+            format!(
+                "{:?}",
+                EmulationError::FaultyKvmExit("FailEntry".to_string())
+            )
+        );
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::InternalError));
+        let res = vcpu.run_emulation();
+        assert!(res.is_err());
+        assert_eq!(
+            format!("{:?}", res.unwrap_err()),
+            format!(
+                "{:?}",
+                EmulationError::FaultyKvmExit("InternalError".to_string())
+            )
+        );
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::SystemEvent(2, 0)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(res.unwrap(), VcpuEmulation::Stopped);
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::SystemEvent(1, 0)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(res.unwrap(), VcpuEmulation::Stopped);
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::SystemEvent(3, 0)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_err());
+        assert_eq!(
+            format!("{:?}", res.unwrap_err()),
+            format!(
+                "{:?}",
+                EmulationError::FaultyKvmExit("SystemEvent(3, 0)".to_string())
+            )
+        );
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Err(errno::Error::new(libc::EAGAIN)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(res.unwrap(), VcpuEmulation::Handled);
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Err(errno::Error::new(libc::ENOSYS)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_err());
+        assert_eq!(
+            format!("{:?}", res.unwrap_err()),
+            format!(
+                "{:?}",
+                EmulationError::FaultyKvmExit("Function not implemented (os error 38)".to_string())
```

Comment:
So I now noticed that the "Received ENOSYS error because KVM failed to emulate an instruction" message will appear in the log files but the returned error message will still be the cryptic one (i.e Function not implemented ...). Shouldn t we also print that as an exit message (i.e return 
```
Err(Error::FaultyKvmExit("Received ENOSYS error because KVM failed to emulate an instruction.".to_string()))
```
?